src/library/scala/concurrent/BlockContext.scala

```diff
@@ -41,40 +41,48 @@ package scala.concurrent
 trait BlockContext {
 
   /** Used internally by the framework;
-   * Designates (and eventually executes) a thunk which potentially blocks the calling `java.lang.Thread`.
-   *
-   * Clients must use `scala.concurrent.blocking` or `scala.concurrent.Await` instead.
-   */
+    * Designates (and eventually executes) a thunk which potentially blocks the calling `java.lang.Thread`.
+    *
+    * Clients must use `scala.concurrent.blocking` or `scala.concurrent.Await` instead.
+    *
+    * In implementations of this method it is RECOMMENDED to first check if `permission` is `null` and
+    * if it is, throw an `IllegalArgumentException`.
+    *
+    * @throws IllegalArgumentException if the `permission` is `null`
+    */
   def blockOn[T](thunk: =>T)(implicit permission: CanAwait): T
 }
 
 object BlockContext {
-  private object DefaultBlockContext extends BlockContext {
+  private final object DefaultBlockContext extends BlockContext {
     override def blockOn[T](thunk: =>T)(implicit permission: CanAwait): T = thunk
   }
 
   /**
-   * @return the `BlockContext` that will be used if no other is found.
-   **/
-  def defaultBlockContext: BlockContext = DefaultBlockContext
+    * The default block context will execute the supplied thunk immediately.
+    * @return the `BlockContext` that will be used if no other is found.
+    **/
+  final def defaultBlockContext: BlockContext = DefaultBlockContext
+
+  private[this] final val contextLocal = new ThreadLocal[BlockContext]()
 
-  private val contextLocal = new ThreadLocal[BlockContext]()
+  private[this] final def prefer(candidate: BlockContext): BlockContext =
+    if (candidate ne null) candidate
+    else {
+      val t = Thread.currentThread
+      if (t.isInstanceOf[BlockContext]) t.asInstanceOf[BlockContext]
+      else DefaultBlockContext
```

Comment:
It's a little depressing that a `match` can't be used because this is faster

Reply:
@viktorklang - Did you actually microbenchmark it?  I generally don't find a substantial difference once the JIT compiler gets through with it.  It's just a store/load instruction pair.  Shouldn't be hard to optimize away.

Reply:
@ichoran Yeah, the difference was noticeable on my machine. However, of course type of machine, type of JVM etc will make it differ. I'm erring on the side of it being easy to make it fast.